Add project roles to JWT token

myieye opened the pull request:
Fixes #6

Adds a user's project roles to their JWT token / user context and uses them to filter for the corresponding projects.

A quick demo (it's not obvious, but the admin is only a member of 3 out of 4 projects):

https://user-images.githubusercontent.com/12587509/214618410-6a80c21f-abfb-43fd-85e0-ee411278632b.mp4

myieye committed:
Disable fixing all .net issues as this is currently a bit out of hand

hahn-kev commented on backend/DataAccess/Entities/User.cs:
```diff
@@ -30,6 +30,8 @@ public User(string role)
         init => _roleString = UserRoleToString(value);
     }
 
+    public required IEnumerable<LfId<Project>> Projects { get; init; }
```
I would reccomend against having fields or properties that are of type IEnumerable (with some exceptions). The reason being that an IEnumerable can be just a list, but it could also be an lazy enumerable, which means each time you access the list you're going to re-execute that enumeration. As an example.
```C#
user.Projects = projectList.Where(p => p.Name == "test")
```
with an IEnumerable that where check isn't executed until you actually iterate the enumerable, but it's executed every time you iterate it. Which means if you write code like this:
```C#
user.Projects.First() == user.Projects.Last();
```
then your `Where` above will be executed twice and the whole list will be iterated twice. This is known as a multiple enumeration issue. To avoid it we should never store an enumerable on a field or property like this, however returning from a method is fine. While this example is fine, you can also get an IEnumerable from the database, and when it get enumerated then the database will execute a query, if you do it multiple times then that means multiple querys which isn't great.

hahn-kev commented on backend/UnitTests/Fixtures/TestWebApplicationFactory.cs:
```diff
@@ -1,27 +1,38 @@
+    }
+
+
+    public new void Dispose()
```
Should try to avoid new, I think in this case we should overload `void Dispose(bool disposing)` and call base passing in the `disposing` variable. Otherwise calling dispose will only working properly if it's called on an instance of `TestWebApplicationFactory` and if that gets cast to an `IDisposable` (maybe by Xunit) then it will only call the base and the mongo runner won't get disposed.

myieye committed:
Teach swagger-UI to accept LfIds as strings

myieye committed:
Enable debugging external code with docs.

myieye committed:
Teach Mongo to serialize LfId dictionary keys as strings

myieye committed:
Downgrade to Linq.V2 for dictionary methods
V3 doesn't seem to support any dictionary methods.
Tests were using V2 already.

myieye committed:
Add project roles to JWT token

myieye committed:
Add JWT token round-trip test

hahn-kev commented on backend/DataAccess/Configuration/LfIdSerializer.cs:
```diff
@@ -149,3 +197,19 @@ public void Serialize(BsonSerializationContext context, BsonSerializationArgs ar
+/// Ensures that our LfIdSerializerProvider is used for all LfId properties including dictionaries
+/// LfIdSerializerProvider.GetSerializer is not called for dictionary members, so we apply it here after each member has been mapped
+/// </summary>
+public class EnsureLfIdMemberSerializationConvention : ConventionBase, IMemberMapConvention
```
I think we can avoid this if we just update `CanConvert` on the `LfIdSerializerProvider`.

myieye replied:
No, `CanConvert` isn't for serializing for Mongo, it's for generating JSON for the web API.

hahn-kev replied:
oh yeah that makes sense.

So I'm not sure if I'm missing something but I just ran a unit test using the debugger and put a breakpoint in `GetSerializer` and it gets called for our dictionary type, so I don't think we need this since it already works. Just try commenting out the code in the `Apply` method.

myieye replied:
You're right 😭 ... I mean, that's good, but some unfortunate combination of events over a prolonged period of time made me think that this wasn't the case...man oh man. .... ☕🏃

hahn-kev commented on backend/DataAccess/DataServiceKernel.cs:
```diff
@@ -18,7 +18,7 @@ public static void Setup(IServiceCollection services)
             var mongoSettings = MongoClientSettings.FromConnectionString(
                 configuration.GetValue<string>("Mongo:ConnectionString"));
-            mongoSettings.LinqProvider = LinqProvider.V3;
+            mongoSettings.LinqProvider = LinqProvider.V2;
```
I think we should move this into a method in the DataAccess project to use as a central config for mongo settings.

hahn-kev commented on backend/UnitTests/Fixtures/IntegrationTestFixture.cs:
```diff
@@ -1,27 +1,40 @@
+
         MongoRunner = SetupMongoRunner();
         var clientSettings = MongoClientSettings.FromConnectionString(MongoRunner.ConnectionString);
+        clientSettings.LinqProvider = LinqProvider.V2;
```
use the central mongo settings config I mentioned previously.

hahn-kev commented on backend/UnitTests/UnitTests.csproj:
```diff
@@ -9,7 +9,8 @@
 
     <ItemGroup>
-        <PackageReference Include="EphemeralMongo6" Version="0.1.3" />
+<PackageReference Include="EphemeralMongo6" Version="0.1.3" />
```
this seems unintentional.

myieye replied:
My formatter did this 😆. I squashed the change, so it's gone now.

hahn-kev commented on backend/DataAccess/Configuration/LfIdSerializer.cs:
```diff
@@ -42,15 +53,42 @@ public override JsonConverter CreateConverter(Type typeToConvert, JsonSerializer
+            IsLfId(i.GetGenericArguments()[0]));
+    }
+
+    private static IBsonSerializer CreateDictionarySerializer<TKey, TValue>() where TKey : notnull
```
nice job, still complicated but at least it's more readable then using reflection.

hahn-kev commented on backend/WebApi/Auth/JwtService.cs:
```diff
@@ -77,8 +80,8 @@ private List<Claim> GetClaims(LfUser user)
             new Claim(JwtRegisteredClaimNames.Sub, user.Id.ToString()),
-            //todo add project specific roles
             new Claim(RoleClaimType, user.Role.ToString()),
+            new Claim(ProjectsClaimType, JsonConvert.SerializeObject(user.Projects)),
```
still need to swap this out for [JsonSerializer](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer?view=net-7.0) from `System.Text.Json`

myieye replied:
Woops 🙂

hahn-kev commented on backend/WebApi/Auth/JwtService.cs:
```diff
@@ -98,7 +101,29 @@ public static TokenValidationParameters TokenValidationParameters(JwtOptions jwt
+
+        var userId = LfId<User>.Parse(idClaim);
+        var userRole = Enum.Parse<UserRole>(roleClaim);
+        var projectRoles = JsonConvert.DeserializeObject<List<UserProjectRole>>(projectRolesClaim) ?? new List<UserProjectRole>();
```
Here too

hahn-kev commented on backend/WebApi/Controllers/TestingController.cs:
```diff
@@ -11,23 +13,34 @@ namespace LanguageForge.WebApi.Controllers;
 
+    [HttpGet("make-jwt-email/{email}")]
+    [AllowAnonymous]
+    public async Task<ActionResult<string>> MakeJwtForUser([EmailAddress] string email)
```
oh nice attribute, does it actually throw if you don't pass in an email?

myieye replied:
Indeed it does 🤓 
```
{
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
  "title": "One or more validation errors occurred.",
  "status": 400,
  "traceId": "00-a320d2c563dc5376c7c82d67db17dd9b-65984fc47359c38d-00",
  "errors": {
    "email": [
      "The email field is not a valid e-mail address."
    ]
  }
}
```

hahn-kev commented on backend/WebApi/Services/UserService.cs:
```diff
@@ -64,12 +64,23 @@ public async Task<bool> DeleteUser(LfId<User> id)
+        }
+
+        var projectRoles = await _systemDbContext.Projects
+            .Find(proj => user.Projects.Contains(proj.Id) && proj.Users.ContainsKey(user.Id))
```
this seems odd, we have to check that user contains project and project contains user. I know that's how the data is structured so we can't change it but I don't like it.

myieye replied:
Yeah. And of course this isn't perfect, because...What if it's only in one list? Who says we should throw it away?

hahn-kev replied:
Yeah, I'll add it as a discussion item for our meeting

hahn-kev commented on backend/WebApi/WebApiKernel.cs:
```diff
@@ -11,15 +10,15 @@ public static void Setup(IServiceCollection services)
-        //todo should be built from JWT
-        services.AddSingleton(new WebUserContext
+        services.AddHttpContextAccessor();
+        services.AddScoped((serviceProvider) =>
```
instead of using a factory function here, maybe we should just inject `IHttpContextAccessor` into `LfWebContext`, just so we don't need to update code here if we need to add something to `LfWebContext`

myieye replied:
I like that idea except for the fact that LfWebContext would then be referencing the JWTService. I'd prefer it to just be given data. But...maybe I'll try it.

myieye replied:
Yeah, I think it looks nice and tidy 🤓, you're right.

hahn-kev replied:
I like it! Now if we really wanted we could do something like pass the active project via a request header and set the active project code in LfWebContext.

hahn-kev commented on backend/WebApi/Auth/JwtService.cs:
```diff
@@ -98,7 +101,29 @@ public static TokenValidationParameters TokenValidationParameters(JwtOptions jwt
+        };
+    }
+
+    public static ILfWebContext BuildUserContext(ClaimsPrincipal user)
```
this should probably only be responsible for returning the `LfUser` instead of the whole Context because this function is in the `JwtService`

myieye replied:
Totally 👍

hahn-kev reviewed (changes requested):
looks good to me. Just a few minor things like the JsonConvert usage and moving away from a factory function to make the WebUserContext.

myieye committed:
Review fixups

myieye commented:
@hahn-kev I've implemeted your suggestions. I accidently squashed the Linq V2 change into the test commit.

hahn-kev reviewed (approved):
looks good to me! I left a couple comments, I still think we can dump the Convention class but I'm fine either way.
